source/common/http/conn_manager_impl.cc

```diff
@@ -661,7 +661,7 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
     // Allow non websocket requests to go through websocket enabled routes.
   }
 
-  if (cached_route_.value()) {
+  if (cached_route_.has_value()) {
```

Comment:
I don't understand why this is a `std::optional<Router:RouteConstSharedPtr>`, but it seems that `Router::RouteMatcher::route` can return a nullptr, in which case I believe `cached_route_.has_value()` can be true with `cached_route_.value()` returning `nullptr`, causing a different crash. 

I think whatever path is crashing never does a route lookup, and so cached_route_ is not initialized to have a value of nullptr. I think it's be worth understanding that better and adding a test that reproduces it.

Assuming this rare no-value case is valid, I think you'd need to do `cached_route_.has_value() && `cached_route_.value()` at these sites.

Note also, there's another `cached_route_.value()` call on line 653.

Reply:
This doesn't seems right, the code asserting `cached_route_` has a value at https://github.com/envoyproxy/envoy/blob/988f7d81f973e17c76db3f900cf145c371d6e8b3/source/common/http/conn_manager_impl.cc#L647

> I think whatever path is crashing never does a route lookup, and so cached_route_ is not initialized to have a value of nullptr. I think it's be worth understanding that better and adding a test that reproduces it.

+1

Reply:
Yes agreed this isn't right. Which line crashes exactly?

Reply:
From looking at the call stack again, I think looks like an issue when we call `createFilterChain()` via an early `sendLocalReply()`. I'm very surprised this path doesn't have test coverage which causes a crash in a test. cc @alyssawilk

Reply:
I'm still trying to suss out when we might sendLocalReply (and do early createFilterChain) before we do decodeHeaders, but still have the stream persist long enough to actually decodeHeaders.   Is it a race onIdletimeout maybe?

Reply:
@alyssawilk I'm probably missing something because a) I've looked at this very quickly and b) I thought we had test coverage of this, but it looks to me like in `decodeHeaders()` we can call `sendLocalReply()` before we call `refreshRouteCache()`. So wouldn't this mean that we would then not have a cached route in these paths and have to guard against it?

Reply:
But in all of these cases when we sendLocalReply we also return, so the route cache should be set.  The only sendLocalReply I could think of which could still result in decodeHeaders is if we had partial headers, then got the rest in the same event loop maybe?  So I added a test for timing out mid-headers (which we should have in any case) and with 5k runs I couldn't trigger any such loop.  I'll probably still land it anyway as it's a pretty reasonable thing to have covered.

However!  After some thought of how else this might happen, if a filter calls clearRouteCache on creation (or rather immediately when it gets the callbacks set, which I think it can) I've got an integration test with a fake filter which does just that and crashes with 
terminate called after throwing an instance of 'absl::bad_optional_access'.  I doubt that's what is happening in prod unless you have a very silly filter, but it is at least a synthetic way of making sure the HCM behaves well when there's no cached route between filter creation and the router filter refreshing it.

It's over here:
https://github.com/envoyproxy/envoy/compare/master...alyssawilk:extra_tests
I'm happy to have it pulled into this PR or land it as a follow-up

Reply:
also to answer your question directly, yes we should totally guard against route cache not being set even if I can't figure out the path which you folks are seeing

Reply:
The call stack is like this:

```
#7  0x00000000007100af in value (this=<optimized out>) at external/com_google_absl/absl/types/optional.h:844
#8  Envoy::Http::ConnectionManagerImpl::ActiveStream::createFilterChain (this=this@entry=0x3cbd2000) at external/envoy/source/common/http/conn_manager_impl.cc:1408
#9  0x000000000071559b in createFilterChain (this=0x3cbd2000) at external/envoy/source/common/http/conn_manager_impl.cc:1001
#10 Envoy::Http::ConnectionManagerImpl::ActiveStream::sendLocalReply(bool, Envoy::Http::Code, std::string const&, std::function<void (Envoy::Http::HeaderMap&)> const&, bool, absl::optional<Envoy::Grpc::Status::GrpcStatus>) (this=this@entry=0x3cbd2000, is_grpc_request=is_grpc_request@entry=false, code=code@entry=Envoy::Http::UpgradeRequired, body=..., modify_headers=..., is_head_request=is_head_request@entry=false, grpc_status=...) at external/envoy/source/common/http/conn_manager_impl.cc:1002
#11 0x000000000071ab1b in Envoy::Http::ConnectionManagerImpl::ActiveStream::decodeHeaders(std::unique_ptr<Envoy::Http::HeaderMap, std::default_delete<Envoy::Http::HeaderMap> >&&, bool) (this=0x3cbd2000, headers=<optimized out>, end_stream=true) at external/envoy/source/common/http/conn_manager_impl.cc:573
```

The crash is inline. I don't think return matters? How does this work when we do an early sendLocalReply() before we eventually populate the route cache? I'm confused how this works or if we don't actually test this case?

Reply:
We totally have tests where we sendLocalReply early, but my guess is we don't have tests where we sendLocalReply early to a request which has an upgrade header, which is one of the places we're checking cached_route_.value() and not cached_route_.has_value() && cached_route_.value() and I think the one in your call stack

I think it'd be safest to always check has_value() and value() in all places.

Reply:
I see. OK. I think it still needs some analysis since I think in some places we do/should assume we have a route, so it might only be in `createFilterChain()` that we need a guard. @tonya11en can you add a test that crashes and then we can discuss a fix?